Support proxy configuration for tasks related functionality

RomanNikitenko opened the pull request:
Signed-off-by: Roman Nikitenko <rnikiten@redhat.com>

### What does this PR do?
Support proxy configuration for tasks related functionality

azatsarynnyy commented:
After merging this PR we need to backport it to master and to 7.15.x

davidfestal commented:
Maybe this code should be extracted in a common component and reused for both the various places where it is used (http client in `che-workspace-client` for example) ? This would avoid some code duplication.
I'm not saying that for this PR of course, but maybe it could be planned for on master ?
wdyt @azatsarynnyy ?

RomanNikitenko commented:
@davidfestal helped me to do testing the changes - thanks a lot!

Looks like the changes fixes the issue.
I'm going to do more testing tomorrow morning.

azatsarynnyy commented:
> Maybe this code should be extracted in a common component and reused for both the various places where it is used (http client in `che-workspace-client` for example) ? This would avoid some code duplication.
> I'm not saying that for this PR of course, but maybe it could be planned for on master ?
> wdyt @azatsarynnyy ?

@davidfestal definitely, it's a good idea. We'll merge the patch into 7.14.x, 7.15.x, and master.
Then we'll create an issue for extracting the proxy related functionality into a separate library/project. To be able to reuse it in both Extensions and Plugins.

tsmaeder commented on plugins/task-plugin/src/machine/websocket.ts:
```diff
@@ -28,22 +31,102 @@ export class ReconnectingWebSocket {
+                    proxy: httpsProxyOptions,
+                    ca: certificateAuthority ? [certificateAuthority] : undefined
+                });
+                const urlIsHttps = (parsedBaseUrl.protocol || 'http:').startsWith('https:');
```
parsedUrl.protocol && parsedUrl.startsWith('https') would be easier to understand, IMO.

tsmaeder commented on plugins/task-plugin/src/machine/websocket.ts:
```diff
@@ -28,22 +31,102 @@ export class ReconnectingWebSocket {
+                const httpOverHttpsAgent = tunnel.httpOverHttps({ proxy: httpsProxyOptions });
+                const httpsOverHttpAgent = tunnel.httpsOverHttp({
+                    proxy: mainProxyOptions,
+                    ca: certificateAuthority ? [certificateAuthority] : undefined
```
If we defined `getCertificateAuthority()` to return `Buffer[] | undefined`, we wouldn't have to write the conditional expression at each use.

tsmaeder commented on plugins/task-plugin/src/machine/websocket.ts:
```diff
@@ -28,22 +31,102 @@ export class ReconnectingWebSocket {
     }
 
+    private getCertificateAuthority(): Buffer | undefined {
+        let certificateAuthority: Buffer | undefined;
```
Together with the comment about the `certificateAuthority` variable above, I would rewrite this method as 
```
    private getCertificateAuthority(): Buffer[] | undefined {
        if (fs.existsSync(SS_CRT_PATH)) {
            return [fs.readFileSync(SS_CRT_PATH)];
        }
        return undefined;
}
```

tsmaeder commented on plugins/task-plugin/src/machine/websocket.ts:
```diff
@@ -28,22 +31,102 @@ export class ReconnectingWebSocket {
+
+    private shouldProxy(hostname: string): boolean {
+        const noProxyEnv = process.env.no_proxy || process.env.NO_PROXY;
+        const noProxy: string[] = noProxyEnv ? noProxyEnv.split(',').map(s => s.trim()) : [];
```
Maybe simpler like this
```
        const noProxyEnv = process.env.no_proxy || process.env.NO_PROXY || '';
        const noProxy: string[] = noProxyEnv.split(',').map(s => s.trim());
```

tsmaeder commented:
If I understand this code correctly, it is using the external URL of the machine exec service. But the machine exec service is running in the same pod, so should be available locally, no? Is there a way we could shortcut going through the proxy?
Also, since we're in a plugin here, why not make executing commands available as part of the che-theia plugin API?
Furthermore, I believe this whole proxying problem would go away if we moved task execution back to being an extension, as is planned, I believe.

RomanNikitenko commented:
> Also, since we're in a plugin here, why not make executing commands available as part of the che-theia plugin API?
Furthermore, I believe this whole proxying problem would go away if we moved task execution back to being an extension, as is planned, I believe.

The best way from my point of view is:
>  if we moved task execution back to being an extension, as is planned
if it solves the problem.

benoitf commented on plugins/task-plugin/src/machine/websocket.ts:
```diff
@@ -28,22 +31,102 @@ export class ReconnectingWebSocket {
+    }
+
+    private shouldProxy(hostname: string): boolean {
+        const noProxyEnv = process.env.no_proxy || process.env.NO_PROXY;
```
out of curiosity, I thought that only `NO_PROXY` is a env variable used, not `no_proxy`

RomanNikitenko replied:
@benoitf I took it [here](https://github.com/eclipse/che-workspace-client/blob/dce065248ae382d7baefd570d9937befc5cdd1c3/src/index.ts#L154)

azatsarynnyy commented:
@RomanNikitenko let's focus on testing and merging this PR ASAP if no critical comments within the reviews.
And address the review comments in the PR https://github.com/eclipse/che-theia/pull/782 for `master` branch.

azatsarynnyy commented:
I've verified it on CRC without a proxy and I see no regression.

Checked:
- running a task from Command Palette
- running a task from My Workspace panel
- terminating the running task
- restarting the already running task

![image](https://user-images.githubusercontent.com/1636395/86230873-601b9700-bb9a-11ea-83f8-bfb9e2569778.png)

RomanNikitenko commented:
Tested for proxy related environment
![tasks_testing_proxy](https://user-images.githubusercontent.com/5676062/86239335-c35ff600-bba7-11ea-83cf-2cc8e5d205b2.gif)

RomanNikitenko commented:
@tsmaeder 
The PR is for `7.14.x` branch.
I'm going to merge the PR as is and apply your comments for `master` branch https://github.com/eclipse/che-theia/pull/782

Is it OK for you?

RomanNikitenko commented:
Tested on https://che.openshift.io/

![openshift_tests_tasks](https://user-images.githubusercontent.com/5676062/86242062-bb568500-bbac-11ea-9ea0-541867f3b9c9.gif)

tsmaeder commented:
Sure.

RomanNikitenko commented:
@tsmaeder 
thank you!
